File: backend/app/routers/reports.py
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException, Query, Request, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from core.limiter import limiter
from db.session import get_db
from models import User
from schemas import (
    CancellationsReportResponse,
    ConsumablesUsageResponse,
    DailyReportResponse,
    FinancialReportResponse,
    MonthlyCheckVsPriceResponse,
    PopularServicesResponse,
    PromoEffectivenessResponse,
    ShiftLoadResponse,
    WasherPayrollResponse,
)
from services.auth_service import get_current_user
from services.excel_export_service import ExcelExportService
from services.reports_service import ReportsService

router = APIRouter(
    prefix="/api/reports",
    tags=["reports"],
)


def _parse_day(date_str: str | None) -> str:
    """Validate and normalize a YYYY-MM-DD date string for the response payload."""
    if not date_str:
        return datetime.now().strftime("%Y-%m-%d")
    try:
        datetime.strptime(date_str, "%Y-%m-%d")
    except ValueError:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST, "Invalid date format. Expected YYYY-MM-DD."
        )
    return date_str


def _parse_report_period(
    date_str: str | None,
) -> tuple[str, date, bool]:
    """Validate and normalize a report period string.

    Supports both YYYY-MM (month) and YYYY-MM-DD (day).
    Returns (payload_date, report_date, is_month).
    """
    if not date_str:
        payload = datetime.now().strftime("%Y-%m")
        return payload, datetime.strptime(payload, "%Y-%m").date(), True
    if len(date_str) == 7:
        try:
            datetime.strptime(date_str, "%Y-%m")
        except ValueError:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST, "Invalid date format. Expected YYYY-MM."
            )
        return date_str, datetime.strptime(date_str, "%Y-%m").date(), True
    if len(date_str) == 10:
        try:
            datetime.strptime(date_str, "%Y-%m-%d")
        except ValueError:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST,
                "Invalid date format. Expected YYYY-MM-DD.",
            )
        return date_str, date.fromisoformat(date_str), False
    raise HTTPException(
        status.HTTP_400_BAD_REQUEST,
        "Invalid date format. Expected YYYY-MM or YYYY-MM-DD.",
    )


@router.get("/monthly-check-vs-price/", response_model=MonthlyCheckVsPriceResponse)
@limiter.limit("30/minute")
async def monthly_report(
    request: Request,
    date_str: str = Query(None, alias="date"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    payload_date, report_date, is_month = _parse_report_period(date_str)
    svc = ReportsService(db)
    return await svc.monthly_report(report_date, payload_date, is_month)


@router.get("/popular-additional-services/", response_model=PopularServicesResponse)
@limiter.limit("30/minute")
async def get_popular_additional_services(
    request: Request,
    date_str: str = Query(None, alias="date"),
    category: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    payload_date, report_date, is_month = _parse_report_period(date_str)
    svc = ReportsService(db)
    return await svc.popular_additional_services(
        report_date, category, payload_date, is_month
    )


@router.get("/consumables-usage/", response_model=ConsumablesUsageResponse)
@limiter.limit("30/minute")
async def get_consumables_usage(
    request: Request,
    date_str: str = Query(None, alias="date"),
    category: str | None = None,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    payload_date, report_date, is_month = _parse_report_period(date_str)
    svc = ReportsService(db)
    return await svc.consumables_usage(
        report_date, category, payload_date, is_month
    )


@router.get("/daily/", response_model=DailyReportResponse)
@limiter.limit("60/minute")
async def daily_report(
    request: Request,
    date_str: str = Query(None, alias="date"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Daily summary: revenue, appointments, top services,
    washers on shift, consumables alerts."""
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    payload_date = _parse_day(date_str)
    svc = ReportsService(db)
    return await svc.daily_report(date.fromisoformat(payload_date), payload_date)


@router.get("/shift-load/", response_model=ShiftLoadResponse)
@limiter.limit("60/minute")
async def shift_load_report(
    request: Request,
    start_date: str,
    end_date: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Weekly shift load analytics (admin only)."""
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов"
        )
    start_date_str = _parse_day(start_date)
    end_date_str = _parse_day(end_date)
    if start_date_str > end_date_str:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "start_date must be before or equal to end_date",
        )
    svc = ReportsService(db)
    return await svc.shift_load_report(
        date.fromisoformat(start_date_str),
        date.fromisoformat(end_date_str),
        start_date_str,
        end_date_str,
    )


@router.get("/financial/", response_model=FinancialReportResponse)
@limiter.limit("60/minute")
async def financial_report(
    request: Request,
    start_date: date,
    end_date: date,
    group_by: str = Query("day", pattern=r"^(day|week|month)$"),
    washer_username: str | None = None,
    wash_type_id: str | None = None,
    promo_id: str | None = None,
    format: str = Query("json", pattern=r"^(json|xlsx)$"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    if start_date > end_date:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "start_date must be before or equal to end_date",
        )
    svc = ReportsService(db)
    data = await svc.financial_report(
        start_date,
        end_date,
        group_by=group_by,
        washer_username=washer_username,
        wash_type_id=wash_type_id,
        promo_id=promo_id,
    )
    if format == "xlsx":
        headers = ["Период", "Записей", "Услуги", "Скидки", "Выручка"]
        rows = [
            [
                i["period"],
                i["appointments_count"],
                i["services_total"],
                i["discounts_total"],
                i["revenue"],
            ]
            for i in data["items"]
        ]
        xlsx, filename = ExcelExportService.generate(
            f"financial_{start_date}_{end_date}", headers, rows
        )
        return Response(
            content=xlsx,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f"attachment; filename={filename}"},
        )
    return data


@router.get("/washer-payroll/", response_model=WasherPayrollResponse)
@limiter.limit("60/minute")
async def washer_payroll_report(
    request: Request,
    start_date: date,
    end_date: date,
    washer_username: str | None = None,
    format: str = Query("json", pattern=r"^(json|xlsx)$"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    if start_date > end_date:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "start_date must be before or equal to end_date",
        )
    svc = ReportsService(db)
    data = await svc.washer_payroll_report(
        start_date, end_date, washer_username
    )
    if format == "xlsx":
        headers = ["Мойщик", "Записей", "Услуги", "Чаевые", "Итого"]
        rows = [
            [
                i["washer_name"],
                i["appointments_count"],
                i["services_total"],
                i["tips_total"],
                i["total"],
            ]
            for i in data["items"]
        ]
        xlsx, filename = ExcelExportService.generate(
            f"washer_payroll_{start_date}_{end_date}", headers, rows
        )
        return Response(
            content=xlsx,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f"attachment; filename={filename}"},
        )
    return data


@router.get("/cancellations/", response_model=CancellationsReportResponse)
@limiter.limit("60/minute")
async def cancellations_report(
    request: Request,
    start_date: date,
    end_date: date,
    reason: str | None = None,
    washer_username: str | None = None,
    wash_type_id: str | None = None,
    format: str = Query("json", pattern=r"^(json|xlsx)$"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    if start_date > end_date:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "start_date must be before or equal to end_date",
        )
    svc = ReportsService(db)
    data = await svc.cancellations_report(
        start_date,
        end_date,
        reason=reason,
        washer_username=washer_username,
        wash_type_id=wash_type_id,
    )
    if format == "xlsx":
        headers = [
            "ID",
            "Дата",
            "Клиент",
            "Авто",
            "Причина",
            "Кем отменено",
            "Потеря",
        ]
        rows = [
            [
                i["appointment_id"],
                i["date"],
                i["client_name"],
                i["car_model"],
                i["reason"],
                i["cancelled_by"],
                i["lost_revenue"],
            ]
            for i in data["items"]
        ]
        xlsx, filename = ExcelExportService.generate(
            f"cancellations_{start_date}_{end_date}", headers, rows
        )
        return Response(
            content=xlsx,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f"attachment; filename={filename}"},
        )
    return data


@router.get("/promo-effectiveness/", response_model=PromoEffectivenessResponse)
@limiter.limit("60/minute")
async def promo_effectiveness_report(
    request: Request,
    start_date: date,
    end_date: date,
    promo_id: str | None = None,
    format: str = Query("json", pattern=r"^(json|xlsx)$"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if current_user.role != "admin":
        raise HTTPException(
            status.HTTP_403_FORBIDDEN, "Доступ только для администраторов."
        )
    if start_date > end_date:
        raise HTTPException(
            status.HTTP_400_BAD_REQUEST,
            "start_date must be before or equal to end_date",
        )
    svc = ReportsService(db)
    data = await svc.promo_effectiveness_report(
        start_date, end_date, promo_id
    )
    if format == "xlsx":
        headers = ["Акция", "Использований", "Выручка", "Скидка"]
        rows = [
            [
                i["promo_name"],
                i["uses_count"],
                i["revenue"],
                i["discount_total"],
            ]
            for i in data["items"]
        ]
        xlsx, filename = ExcelExportService.generate(
            f"promo_effectiveness_{start_date}_{end_date}", headers, rows
        )
        return Response(
            content=xlsx,
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": f"attachment; filename={filename}"},
        )
    return data

File: telegram-miniapp/src/services/reports.ts
import { api } from './api'

export interface DailyReport {
  report_date: string
  revenue: number
  appointments_count: number
  completed_count: number
  average_check: number
  box_occupancy: Record<string, number>
  top_services: Array<{
    name: string
    count: number
    revenue: number
  }>
  washers_on_shift: number
  consumables_alert: string[]
}

export interface FinancialReport {
  summary: Record<string, number>
  items: Array<{
    period: string
    appointments_count: number
    services_total: number
    discounts_total: number
    revenue: number
  }>
}

export interface WasherPayrollReport {
  items: Array<{
    washer_username: string
    washer_name: string
    appointments_count: number
    services_total: number
    tips_total: number
    total: number
  }>
}

export interface PromoEffectivenessReport {
  items: Array<{
    promo_id: string | null
    promo_name: string
    uses_count: number
    revenue: number
    discount_total: number
  }>
}

export interface CancellationsReport {
  summary: Record<string, number>
  items: Array<{
    appointment_id: string
    date: string
    client_name: string
    car_model: string
    reason: string | null
    cancelled_by: string
    lost_revenue: number
  }>
}

export interface PopularServicesReport {
  month: string
  category: string | null
  items: Array<{
    name: string
    count: number
    category: string | null
  }>
}

export interface MonthlyCheckVsPriceReport {
  month: string
  items: Array<{
    car_model: string
    avg_check: number
    visit_count: number
  }>
}

export interface ConsumablesUsageReport {
  month: string
  category: string | null
  items: Array<{
    consumable_name: string
    unit: string
    total_used: number
  }>
}

export interface ShiftLoadReport {
  startDate: string
  endDate: string
  targetWeeklyMinutesPerWasher: number
  dailyHours: Array<{
    date: string
    confirmedMinutes: number
    pendingMinutes: number
  }>
  washerStats: Array<{
    userId: number
    displayName: string
    confirmedMinutes: number
  }>
  statusCounts: Record<string, number>
  conflictCount: number
  availabilityCoverage: Record<string, number>
}

export async function getDailyReport(date: string, signal?: AbortSignal): Promise<DailyReport> {
  const res = await api.get('/reports/daily/', { params: { date }, signal })
  return res.data
}

export async function getFinancialReport(
  startDate: string,
  endDate: string,
  groupBy: 'day' | 'week' | 'month' = 'day',
  signal?: AbortSignal,
): Promise<FinancialReport> {
  const res = await api.get('/reports/financial/', {
    params: { start_date: startDate, end_date: endDate, group_by: groupBy },
    signal,
  })
  return res.data
}

export async function getWasherPayrollReport(
  startDate: string,
  endDate: string,
  washerUsername?: string,
  signal?: AbortSignal,
): Promise<WasherPayrollReport> {
  const res = await api.get('/reports/washer-payroll/', {
    params: { start_date: startDate, end_date: endDate, washer_username: washerUsername },
    signal,
  })
  return res.data
}

export async function getPromoEffectivenessReport(
  startDate: string,
  endDate: string,
  promoId?: string,
  signal?: AbortSignal,
): Promise<PromoEffectivenessReport> {
  const res = await api.get('/reports/promo-effectiveness/', {
    params: { start_date: startDate, end_date: endDate, promo_id: promoId },
    signal,
  })
  return res.data
}

export async function getCancellationsReport(
  startDate: string,
  endDate: string,
  params: {
    reason?: string
    washer_username?: string
    wash_type_id?: string
  } = {},
  signal?: AbortSignal,
): Promise<CancellationsReport> {
  const res = await api.get('/reports/cancellations/', {
    params: { start_date: startDate, end_date: endDate, ...params },
    signal,
  })
  return res.data
}

export async function getPopularAdditionalServicesReport(
  date: string,
  category?: string,
  signal?: AbortSignal,
): Promise<PopularServicesReport> {
  const res = await api.get('/reports/popular-additional-services/', {
    params: { date, category },
    signal,
  })
  return res.data
}

export async function getMonthlyCheckVsPriceReport(
  date: string,
  signal?: AbortSignal,
): Promise<MonthlyCheckVsPriceReport> {
  const res = await api.get('/reports/monthly-check-vs-price/', { params: { date }, signal })
  return res.data
}

export async function getConsumablesUsageReport(
  date: string,
  category?: string,
  signal?: AbortSignal,
): Promise<ConsumablesUsageReport> {
  const res = await api.get('/reports/consumables-usage/', {
    params: { date, category },
    signal,
  })
  return res.data
}

export async function getShiftLoadReport(
  startDate: string,
  endDate: string,
  targetWeeklyMinutes?: number,
  signal?: AbortSignal,
): Promise<ShiftLoadReport> {
  const res = await api.get('/reports/shift-load/', {
    params: {
      start_date: startDate,
      end_date: endDate,
      target_weekly_minutes: targetWeeklyMinutes,
    },
    signal,
  })
  return res.data
}
